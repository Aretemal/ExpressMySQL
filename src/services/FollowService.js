import { Op } from 'sequelize';
import Follow from '../models/follow.js';
import User from '../models/user.js';

class FollowService {
  async follow(firstUserId, secondUserId) {
    const connection = Follow.create({
      followerId: firstUserId, followingId: secondUserId,
    });
    return connection;
  }

  async approve(firstUserId, secondUserId, next) {
    const approvedFollow = await Follow.findOne({
      where: {
        followerId: secondUserId,
        followingId: firstUserId,
      },
    });
    if (!approvedFollow) {
      next({ errorsArray: [{ msg: '440' }], title: 'Service Error' });
      return;
    }
    await approvedFollow.update({ approvedAt: (new Date()) });
    return approvedFollow;
  }

  async unfollow(unSubscriberId, userId) {
    const deletedConnection = await Follow.findOne({
      where: {
        followerId: userId,
        followingId: unSubscriberId,
      },
    });
    if (!deletedConnection) {
      const deletedConnection = await Follow.findOne({
        where: {
          followerId: unSubscriberId,
          followingId: userId,
        },
      });
      if (deletedConnection.approvedAt === null) {
        await deletedConnection.destroy();
        return null;
      }
      await deletedConnection.update({
        followerId: userId,
        followingId: unSubscriberId,
        approvedAt: null,
      });
      return deletedConnection;
    }
    await deletedConnection.update({ approvedAt: null });
    return deletedConnection;
  }

  async getFriends(id) {
    const connections = await Follow.findAll({
      include: [{
        model: User,
        association: 'following',
      }, {
        model: User,
        association: 'follower',
      }],
      where: {
        [Op.or]: [
          { followerId: id, approvedAt: { [Op.ne]: null } },
          { followingId: id, approvedAt: { [Op.ne]: null } },
        ],
      },
    });
    const friends = connections.map((item) => {
      if (item.dataValues.follower.dataValues.id === id) {
        return item.dataValues.following.dataValues;
      }
      return item.dataValues.follower.dataValues;
    });
    return friends;
  }

  async getSubscriptions(id) {
    const connections = await Follow.findAll({
      include: {
        model: User,
        association: 'following',
      },
      where: {
        followerId: id,
        approvedAt: null,
      },
    });
    const subscriptions = connections.map((item) => item.dataValues.following.dataValues);
    return subscriptions;
  }

  async getSubscribers(id) {
    const connections = await Follow.findAll({
      include: {
        model: User,
        association: 'follower',
      },
      where: {
        followingId: id,
        approvedAt: null,
      },
    });
    const subscribers = connections.map((item) => item.dataValues.follower.dataValues);
    return subscribers;
  }
}
export default new FollowService();
