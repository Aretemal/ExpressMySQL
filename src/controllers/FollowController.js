import { validationResult } from 'express-validator';
import FollowService from '../services/FollowService.js';
import FollowSerializer from '../serializers/FollowSerializer.js';

class FollowController {
  async follow(req, res, next) {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }
    const { id } = req.body;
    const follow = await FollowService.follow(req.user.id, id);
    req.serializer = new FollowSerializer(follow);
    next();
  }

  async approve(req, res, next) {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }
    const { id } = req.body;
    const follow = await FollowService.approve(req.user.id, id);
    req.serializer = new FollowSerializer(follow);
    next();
  }

  async unfollow(req, res, next) {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }
    const { id } = req.body;
    const follow = await FollowService.unfollow(req.user.id, id);
    req.serializer = new FollowSerializer(follow);
    next();
  }
}
export default new FollowController();
