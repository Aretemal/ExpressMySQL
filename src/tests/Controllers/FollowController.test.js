/* eslint-disable no-undef */
// NODE_OPTIONS=--experimental-vm-modules npx jest
import sinon from 'sinon';
import FollowController from '../../controllers/FollowController';
import FollowService from '../../services/FollowService';
import { req, res } from './ReqRes.js';

describe('Follow Controller : ', () => {
  test('Creates a new connection', async () => {
    const stub = sinon.stub(FollowService, 'follow');
    stub.returns({ followerId: 1, followingId: 2, approvedAt: 0 });
    await FollowController.follow(req, res);
    expect(res.dataJS.data.attributes.attributes.followerId).toBe(1);
    expect(res.dataJS.data.attributes.attributes.followingId).toBe(2);
    expect(res.dataJS.data.attributes.attributes.approvedAt).toBe(0);
  });
  test('Approve connection', async () => {
    const stub = sinon.stub(FollowService, 'approve');
    const timeD = new Date();
    stub.returns({ followerId: 1, followingId: 2, approvedAt: timeD });
    await FollowController.approve(req, res);
    expect(res.dataJS.data.attributes.attributes.followerId).toBe(1);
    expect(res.dataJS.data.attributes.attributes.followingId).toBe(2);
    expect(res.dataJS.data.attributes.attributes.approvedAt).toBe(timeD);
  });
  test('Unfollow', async () => {
    const stub = sinon.stub(FollowService, 'unfollow');
    stub.returns({ followerId: 1, followingId: 2, approvedAt: 0 });
    await FollowController.unfollow(req, res);
    expect(res.dataJS.data.attributes.attributes.followerId).toBe(1);
    expect(res.dataJS.data.attributes.attributes.followingId).toBe(2);
    expect(res.dataJS.data.attributes.attributes.approvedAt).toBe(0);
  });
});
