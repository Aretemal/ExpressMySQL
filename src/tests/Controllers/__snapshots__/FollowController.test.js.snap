// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Follow Controller : Approve should changed connection 1`] = `
FollowSerializer {
  "resource": {
    "approvedAt": 0,
    "followerId": 1,
    "followingId": 2,
  },
}
`;

exports[`Follow Controller : Follow should create and return a new connection 1`] = `
FollowSerializer {
  "resource": {
    "approvedAt": 0,
    "followerId": 1,
    "followingId": 2,
  },
}
`;

exports[`Follow Controller : Unfollow should return deleted connection 1`] = `
FollowSerializer {
  "resource": {
    "approvedAt": 0,
    "followerId": 1,
    "followingId": 2,
  },
}
`;

exports[`Follow Controller : getFriends should return array of users 1`] = `
CollectionSerializer {
  "resource": [
    {
      "ava": "example",
      "email": "example",
      "firstName": "example",
      "id": 1,
      "lastName": "example",
      "login": "example",
      "status": "example",
    },
  ],
}
`;

exports[`Follow Controller : getSubscribers should return array of users 1`] = `
CollectionSerializer {
  "resource": [
    {
      "ava": "example",
      "email": "example",
      "firstName": "example",
      "id": 1,
      "lastName": "example",
      "login": "example",
      "status": "example",
    },
  ],
}
`;

exports[`Follow Controller : getSubscriptions should return array of users 1`] = `
CollectionSerializer {
  "resource": [
    {
      "ava": "example",
      "email": "example",
      "firstName": "example",
      "id": 1,
      "lastName": "example",
      "login": "example",
      "status": "example",
    },
  ],
}
`;
